src/codecs/jxl/decoder.rs

```diff
@@ -0,0 +1,163 @@
+            ))
+        })?;
+        let mut keyframes = Vec::new();
+        let mut renderer = image.renderer();
+        loop {
+            let result = renderer.render_next_frame().map_err(|e| {
+                ImageError::Decoding(DecodingError::new(ImageFormat::Jxl.into(), e))
+            })?;
+            match result {
+                jxl_oxide::RenderResult::Done(frame) => keyframes.push(frame),
+                jxl_oxide::RenderResult::NeedMoreData => {
+                    return Err(ImageError::Decoding(DecodingError::new(
+                        ImageFormat::Jxl.into(),
+                        "Unexpected end of file",
+                    )))
+                }
+                jxl_oxide::RenderResult::NoMoreFrames => break,
+            }
+        }
+        let pixfmt = renderer.pixel_format();
+        let metadata = &image.image_header().metadata;
+        let bits_per_sample = metadata.bit_depth.bits_per_sample();
+        let bitdepth = BitDepth::new(bits_per_sample);
+
+
+        let colortype = match (pixfmt, bitdepth) {
+            (PixelFormat::Gray, BitDepth::Eight) => ColorType::L8,
+            (PixelFormat::Gray, BitDepth::Sixteen) => ColorType::L16,
+            //
+            (PixelFormat::Graya, BitDepth::Eight) => ColorType::La8,
+            (PixelFormat::Graya, BitDepth::Sixteen) => ColorType::La16,
+            //
+            (PixelFormat::Rgb, BitDepth::Eight) => ColorType::Rgb8,
+            (PixelFormat::Rgb, BitDepth::Sixteen) => ColorType::Rgb16,
+            (PixelFormat::Rgb, BitDepth::ThirtyTwo) => ColorType::Rgb32F,
+            //
+            (PixelFormat::Rgba, BitDepth::Eight) => ColorType::Rgba8,
+            (PixelFormat::Rgba, BitDepth::Sixteen) => ColorType::Rgba16,
+            (PixelFormat::Rgba, BitDepth::ThirtyTwo) => ColorType::Rgba32F,
+            //
+            _ => {
+                return Err(unsupported_color(ExtendedColorType::Unknown(
+                    bits_per_sample as u8,
+                )))
+            }
+        };
+
+        Ok(Self {
+            image,
+            bitdepth: BitDepth::new(bits_per_sample),
+            colortype,
+            keyframes,
+        })
+    }
+}
+
+fn unsupported_color(ect: ExtendedColorType) -> ImageError {
+    ImageError::Unsupported(UnsupportedError::from_format_and_kind(
+        ImageFormat::Jxl.into(),
+        UnsupportedErrorKind::Color(ect),
+    ))
+}
+
+impl<'a, R: 'a + Read> ImageDecoder<'a> for JxlDecoder<R> {
+    type Reader = Cursor<Vec<u8>>;
+
+    fn dimensions(&self) -> (u32, u32) {
+        let size = &self.image.image_header().size;
+        (size.width, size.height)
+    }
+
+    fn color_type(&self) -> ColorType {
+        self.colortype
+    }
+
+    fn into_reader(self) -> ImageResult<Self::Reader> {
+            let data = decoder_to_vec(self)?;
+            Ok(Cursor::new(data))
+    }
+
+    fn icc_profile(&mut self) -> Option<Vec<u8>> {
+        self.image.embedded_icc().map(Vec::from)
```

Comment:
Decoded images are in `decoded_icc()` profile, but it seems like `JxlImage` lacks the method. I'll fix that in the next release of jxl-oxide.